### ISSUE TITLE
Bignum: improve modulus life cycle

### ISSUE THREAD

**Issue by yanesca** [Mbed-TLS/mbedtls#6757]
### Suggested enhancement

Extend the modulus life cycle API to make memory management safer.

### Justification

Currently `mbedtls_mpi_mod_modulus_free` doesn't actually free the limbs of the modulus, only the rest of the structure. There's a widespread precedent in the library that xxx_free() doesn't free its argument (which is a gotcha in itself), but frees all embedded resources. A modulus structure is unusual in that it doesn't own the limbs of the modulus.

(See https://github.com/Mbed-TLS/mbedtls/pull/6747/files#r1043716624)

**Comment by yanesca** [comment]
**Proposal 1:** Add a parameter to setup (and a field to the struct) that indicates the ownership of the modulus limbs. Free then would free or not free the modulus limbs based on that field.

**Proposal 2:** Have two setup functions one of them handing over the ownership of the limbs to the modulus structure and the other doesn't.

**Comment by gilles-peskine-arm** [comment]
I think we've had good success with Mbed TLS's general philosophy of memory ownership: when structure contains a pointer, it owns the pointer. This leads to copying data even in cases where we could get away with multiple references to the same memory, hence increased memory usage. But it helps a lot to avoid memory management errors. We still have memory management errors occasionally, but it's due to having not-so-consistent handling locally. We have leaks when we accidentally create error paths that skip cleanup, or use-after-free when we accidentally keep a local reference to a freed block. Sometimes we make a mistake as to _when_ to clean up, but we don't really run into situations where it's hard to tell _what_ to clean up.

So I would like to keep the same philosophy in the new bignum as much as is practical.

For modulus structures, what are the use cases for non-ownership? The one case I'm aware of is ECC where we'd like the modulus to be just one compile-time constant. Are there others? If it's just that in ECC, then I think `bignum_mod` should only expose an interface where the modulus structure owns the limb array, and the relevant parts of ECC should use separate auxiliary functions from `bignum_mod_raw`. And in `bignum_mod` there should be “normal” setup/free functions where pointer ownership is not transferred.

For residue structures, I recognize that the most common usage will be that the residue is a temporary reference to a segment of some long-lived limb array, so the current setup/release interface is probably the right one for most uses. Unit tests and rather the exception, there I think we'll often want to allocate memory for a residue, so we might have test helper functions for that.

**Comment by tom-cosgrove-arm** [comment]
Do we want a clear statement like this for residue?

```
diff --git a/library/bignum_mod.h b/library/bignum_mod.h
index 0a8f4d3d0..41a57c2dc 100644
--- a/library/bignum_mod.h
+++ b/library/bignum_mod.h
@@ -67,40 +67,44 @@ typedef struct {
:
 /** Setup a residue structure.
  *
  * The residue will be set up with the buffer \p p and modulus \p m.
  *
  * The memory pointed to by \p p will be used by the resulting residue structure.
  * The value at the pointed-to memory will be the initial value of \p r and must
  * hold a value that is less than the modulus. This value will be used as-is
  * and interpreted according to the value of the `m->int_rep` field.
  *
  * The modulus \p m will be the modulus associated with \p r. The residue \p r
  * should only be used in operations where the modulus is \p m.
  *
+ * Unlike most `mbedtls_*_setup()` functions, `mbedtls_mpi_mod_residue_setup()`
+ * does not take ownership of the data \p p passed in, and so there is no
+ * corresponding `free()` function, but `mbedtls_mpi_mod_residue_release()`.
+ *
  * \param[out] r    The address of the residue to setup.
  * \param[in] m     The address of the modulus related to \p r.
  * \param[in] p     The address of the limb array containing the value of \p r.
  *                  The memory pointed to by \p p will be used by \p r and must
  *                  not be modified in any way until after
  *                  mbedtls_mpi_mod_residue_release() is called. The data
  *                  pointed to by \p p must be less than the modulus (the value
  *                  pointed to by `m->p`) and already in the representation
  *                  indicated by `m->int_rep`.
  * \param p_limbs   The number of limbs of \p p. Must be the same as the number
  *                  of limbs in the modulus \p m.
  *
  * \return      \c 0 if successful.
  * \return      #MBEDTLS_ERR_MPI_BAD_INPUT_DATA if \p p_limbs is less than the
  *              limbs in \p m or if \p p is not less than \p m.
  */
 int mbedtls_mpi_mod_residue_setup( mbedtls_mpi_mod_residue *r,
                                    const mbedtls_mpi_mod_modulus *m,
                                    mbedtls_mpi_uint *p,
                                    size_t p_limbs );
```

**Comment by yanesca** [comment]
> For modulus structures, what are the use cases for non-ownership?

As you mentioned, setting up moduli from compile-time constants. For now, the scope is ECC and this is the only use case we have. We don't have a use case where we would want to transfer ownership.

Thinking forward to RSA, we indeed could clone the input MPI (potentially increasing peak memory consumption, which should be fine as the caller can free the original array immediately after setup).

> For residue structures, I recognize that the most common usage will be that the residue is a temporary reference to a segment of some long-lived limb array

No, we mostly would do that in the ECP module and there we work with raw pointers. The use case for non-ownership is setting up the EC base point from compile-time constants. I guess the same approach (the one you suggested for moduli) could work here too. My only concern is that, it would add code using residue structs to mod_raw, which I was reluctant to do (as using/not using the residue structure was a clear conceptual line between mod and mod_raw, but I guess it is worth to blur the line to make memory management more robust).

**Comment by yanesca** [comment]
> Do we want a clear statement like this for residue?

Yes, we probably should, but since we are likely to change it relatively soon, it is probably best to wait with that.

**Comment by yanesca** [comment]
The modulus is constant during its lifetime and having the limbs in a compile time constant is safe and straightforward. Residue structures however often change their value. If we set it up from a compile time constant directly, it is easier to accidentally try writing to it.

Should we do something about that? Should we keep track the storage location in the structure and error out when trying to write to it? (This wouldn't be much of a help as we can't do it in ECP, where raw pointers are used.) Should we just clone them and accept the cost in RAM? (It should be relatively low in most cases, but can be more significant when it comes to multithreading.)

**Comment by gilles-peskine-arm** [comment]
> If we set it up from a compile time constant directly, it is easier to accidentally try writing to it.

Where would we use a compile-time constant with the mod (not mod_raw) interface?

However, I think at some point we'll want to use mod with a big temporary in RSA. So I do think we'll run into that use case for a residue structure that doesn't own the limb array.

**Comment by yanesca** [comment]
> Where would we use a compile-time constant with the mod (not mod_raw) interface?

For now the compile-time constants in residues are coordinates of a point which are (and should) only used in point operations which are (and should be) in the ecp module, which uses the mod_raw interface.
